### ISSUE TITLE
NNBD introduced breaking change to Isolate.spawn functions

### ISSUE THREAD

**Issue by gaaclarke** [dart-lang/sdk#44181]
Projects that executed fine have been broken with the introduction of NNBD without any changes.  This appears to happen because `Isolate.spawnUri` doesn't inherit the settings from the calling context so a non-NNBD context spawning an isolate will attempt to make a NNBD isolate unless the called code has a language version annotation.

Examples:
- https://github.com/flutter/flutter/issues/70355
- https://cirrus-ci.com/task/5943556349100032

Expected result:
I expect Isolate.spawn functions to default the spawned isolate to the same settings as the spawning isolate unless otherwise specified.

Dart version:
2.12.0 (build 2.12.0-31.0.dev)

**Comment by alexmarkov** [comment]
Related: https://github.com/dart-lang/sdk/issues/40632, https://github.com/dart-lang/sdk/issues/41207.
@jakemac53 @natebosch @leafpetersen @lrhn

**Comment by lrhn** [comment]
If I read this correctly, it's `Isolate.spawnUri` which opts its spawned isolate into sound null safety based on the entry-point library.
This is dangerous if the library may depend on non-null-safe libraries, and there is no way to override the behavior.
(Or rather, one will have to create a non-null-safe entry point which forwards to the real entry-point library, and use that until all dependencies have been migrated).

It's only a problem for out-of-order migration.

**Comment by natebosch** [comment]
> It's only a problem for out-of-order migration.

The problem we are hitting is that it's a problem for users who upgrade their SDK but don't care about null safety yet.

The reasoning to default to opt-in if the VM doesn't know the language version is the same as the reasoning to default the overall VM to opt-in if you run a script that isn't in a package - we want the default to be null safety because it's the best default in the long term. In the short term during the SDK switch it does cause some pain.

**Comment by gaaclarke** [comment]
> If I read this correctly, it's `Isolate.spawnUri` which opts its spawned isolate into sound null safety based on the entry-point library.

With the Pigeon issue mentioned in the description, the spawning context was non-NNBD but it was attempting to create a NNBD isolate.  The code worked fine until there was an upgrade of Dart and it stopped working.  I had to fix it by adding a language annotation to the spawned code.  

Something like this this could would be broken with a Dart upgrade (see pigeon for a real example):

```dart
void main() {
  final String code = """
void main(List<String> args, SendPort sendPort) async {
  print('hello');
}
""";

  String path = tempFilePath();
  File(path).writeAsString(code);
  Isolate.spawnUri(Uri.file(path), null, null);
}
```

**Comment by jakemac53** [comment]
Yes this is a breaking change but it is exactly the same breaking change that is being made for all dart code.

If you have some code which is not a part of a package (including raw code in a string), it will be opted in to null safety. That is the default language version.

It is imo correct for both `Isolate.spawnUri(<uri>)` and `dart <uri>` to have the same behavior, and they do today, which is what you are seeing.

We should likely call this out explicitly in the breaking change announcements for NNBD though, if it isn't already.

**Comment by mit-mit** [comment]
> We should likely call this out explicitly in the breaking change announcements for NNBD though, if it isn't already.

@leafpetersen have we actually sent a general breaking change notification for null safety overall? With it going to beta, and `dart <file name>` getting null safety by-default, that might be a good idea.

**Comment by leafpetersen** [comment]
> @leafpetersen have we actually sent a general breaking change notification for null safety overall?

No.  I can do so.

**Comment by nex3** [comment]
When you first announced null safety, you made a big point of saying ["Null safety is backwards compatible"](https://medium.com/dartlang/announcing-sound-null-safety-defd2216a6f3#aea6). This is a clear backwards incompatibility, and one that's already affecting real users—Dart Sass's CI has been broken for six days because of it, and we have no recourse but to wait for an upstream fix. And that's not counting the numerous changes we've had to make to a number of other packages we use to work around the underlying VM-default issue.

I understand the desire to move everyone onto null safety, and to have the VM default to the language version that's ultimately going to be universal in the future. But you can't just go about making breaking changes left and right, *especially* when you've publicly promised not to do so. Just because the new behavior is better doesn't negate the pain, frustration, and lost productivity that those breakages cause.

In Sass, when we need to make a breaking change, we follow a process for doing so that's designed to minimize user pain. First, we mark the old behavior as deprecated, and produce some kind of message or warning when it happens. (This is also how, for example, what Node.js did when they transitioned unhandled promises from being silently ignored to crashing the program.) We allow that deprecation to exist in the ecosystem for a substantial length of time—usually a year or more, but sometimes less if the breaking change is urgently necessary to support a CSS feature—and only once we're confident everyone has had a chance to migrate away from the old behavior do we add the new behavior.

I would suggest a similar pattern here. In Dart 2.12, continue to treat unmarked scripts as being the lowest possible language version, but if they have null-unsafe dependencies produce a warning message that indicates that their version should be explicitly marked. Then in Dart 2.15 or whatever, once null safety is more or less universal, switch the default with way lower risk of breakage.

**Comment by gaaclarke** [comment]
@nex3 I agree Dart should have avoided this breaking change.  Seems like they have enough contextual information to move people to NNBD and not break people in this case, but in the meantime I was able to work around the issues by:
1) Adding language specifications to the top of my scripts
2) Upgrading the `test` package

Hopefully you can do that, too, to unblock your CI.

**Comment by franklinyow** [comment]
@leafpetersen I'm assigning this to you, please assign a priority, Thx!

**Comment by jathak** [comment]
Any updates on this? Since 2.12 was released to stable, this is now a breaking change. [Dart Sass's CI is still broken](https://github.com/sass/dart-sass/pull/1247/checks?check_run_id=2034298372), which is forcing us to pin to 2.11.

**Comment by lrhn** [comment]
Could we change `Isolate.spawnUri` to always make the new isolate non-sound when the spawning isolate is non-sound?
That would mean that you cannot spawn a sound isolate ever, if your starting isolate is unsound, but when your entire program is sound, you will get sound isolates too.

**Comment by jakemac53** [comment]
> Could we change `Isolate.spawnUri` to always make the new isolate non-sound when the spawning isolate is non-sound?
> That would mean that you cannot spawn a sound isolate ever, if your starting isolate is unsound, but when your entire program is sound, you will get sound isolates too.

This would mean you couldn't run sound tests at all until the test runner is migrated, etc. At this point it is _nearly_ released as a sound mode app but flutter users won't get that until the next flutter release due to package pinning, so no flutter tests would be able to run in sound mode.

It would also introduce a weird divergence in behavior between isolates and normal apps which I don't think is justified, and isn't the desired behavior either.

**Comment by nex3** [comment]
This isn't the first time we've run into the broad category of issues where "sometimes you want an isolate to inherit X from its parent and sometimes you want to control X explicitly". Maybe there's a need for an opaque `Map<Symbol, Object> flags` parameter that can be used to trigger migration-compatibility behavior in a way that won't introduce API bloat once a migration is completed? For example, `Isolate.spawnUri()` could default to the same soundness as its parent, but `test` could pass in `flags: {#forceSoundNullSafety: true}` when it wants to ensure tests run in null-safe mode. Then when null safety is no longer optional, `dart:isolate` could just ignore `#forceSoundNullSafety` going forward (or similar for any future migration).
